### PR TITLE
Support NG Monitoring

### PR DESCRIPTION
**Pull request by KanShiori**
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support to deploy TiDBNGMonitoring
```

### CODE REVIEW

File: pkg/manager/member/utils.go [hunk 1]
```diff
@@ -74,30 +73,6 @@ func annotationsMountVolume() (corev1.VolumeMount, corev1.Volume) {
 	return m, v
 }
 
-// statefulSetIsUpgrading confirms whether the statefulSet is upgrading phase
-func statefulSetIsUpgrading(set *apps.StatefulSet) bool {
```
Comment (KanShiori) [inline pkg/manager/member/utils.go:78]: move some function to a public pkg `pkg/manager/utils` in order to reuse

File: pkg/apis/pingcap/v1alpha1/tidbngmonitoring_types.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	// Clusters reference TiDB cluster
+	//
+	// +kubebuilder:validation:MaxItems=1
+	// +kubebuilder:validation:MinItems=1
+	Clusters []TidbClusterRef `json:"clusters"`
+
+	// Paused pause controller if it is true
+	Paused bool `json:"paused,omitempty"`
+
+	// Persistent volume reclaim policy applied to the PVs that consumed by TiDB cluster
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/tidbngmonitoring_types.go:75]: `by TiDB cluster` need to update?

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+
+	return nil
+}
+
+func (m *ngMonitoringManager) syncService(tngm *v1alpha1.TiDBNGMonitoring) error {
+	ns := tngm.GetNamespace()
+	name := tngm.GetName()
+
+	if tngm.Spec.Paused {
+		klog.V(4).Infof("tikv cluster %s/%s is paused, skip syncing for ng monitoring headless service", ns, name)
```
Comment (csuzhangxc) [inline pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:76]: not `tikv`

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+	}
+	isOrphan := metav1.GetControllerOf(oldSvc) == nil
+	if !equal || isOrphan {
+		svc := *oldSvc
+		svc.Spec = newSvc.Spec
+		err = controller.SetServiceLastAppliedConfigAnnotation(&svc)
+		if err != nil {
+			return err
+		}
+		// Adopt headless-service created by helm
```
Comment (csuzhangxc) [inline pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:110]: ? can we remove the handler of `isOrphan` for ngm?

File: pkg/apis/pingcap/v1alpha1/crd_kinds.go [hunk 5]
```diff
@@ -53,6 +53,10 @@ const (
 	TidbClusterAutoScalerKind    = "TidbClusterAutoScaler"
 	TidbClusterAutoScalerKindKey = "tidbclusterautoscaler"
 
+	TiDBNGMonitoringName    = "tidbngmonitorings"
+	TiDBNGMonitoringKind    = "TiDBNGMonitoring"
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/crd_kinds.go:57]: name is a hard thing 😢

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go (hunk 4 above)
Comment (KanShiori) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go]: removed

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go (hunk 3 above)
Comment (KanShiori) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go]: done

File: pkg/apis/pingcap/v1alpha1/tidbngmonitoring_types.go (hunk 2 above)
Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/tidbngmonitoring_types.go]: done

File: pkg/apis/pingcap/v1alpha1/crd_kinds.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/crd_kinds.go]: done

File: pkg/apis/pingcap/v1alpha1/tidbngmonitoring_types.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/pingcap/tidb-operator/pkg/apis/util/config"
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TidbNGMonitoring contains the spec and status of tidb ng monitor
+//
+// +genclient
+// +genclient:noStatus
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbngmonitoring_types.go:26]: We can enable status subresource for ngmonitoring.

File: pkg/apis/pingcap/v1alpha1/validation/validation.go [hunk 7]
```diff
@@ -62,6 +62,21 @@ func ValidateDMCluster(dc *v1alpha1.DMCluster) field.ErrorList {
 	return allErrs
 }
 
+// ValidateTiDBNGMonitoring validates a TidbNGMonitoring
+func ValidateTiDBNGMonitoring(tngm *v1alpha1.TidbNGMonitoring) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// medata
+	fldPath := field.NewPath("metadata")
+	allErrs = append(allErrs, validateAnnotations(tngm.ObjectMeta.Annotations, fldPath.Child("annotations"))...)
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/validation/validation.go:71]: No need to validate this as it's for the TiDB cluster components only.

**Review by csuzhangxc — APPROVED** [review]

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog"
+)
+
+const (
+	ngmPodDataVolumeMountDir   = "/var/lib/ng-monitoring" // the mount path for ng monitoring data volume
+	ngmPodConfigVolumeMountDir = "/etc/ng-monitoring"     // the dir for ng monitoring config
+	ngmPodConfigFilename       = "ng-monitoring.toml"     // the filename of config file
+
+	ngmConfigMapConfigKey = "ng-monitoring-config" // the key for config data in config map
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:39]: yes, but we can't distinguish config if all components use the same key in `updateConfigMap`, I don't know if we will need to distinguish in the future

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+			return false, nil
+		}
+		if revisionHash != oldSts.Status.UpdateRevision {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func GenerateNGMonitoringStatefulSet(tngm *v1alpha1.TidbNGMonitoring, cm *corev1.ConfigMap) (*apps.StatefulSet, error) {
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:241]: not necessary, but the function doesn't have any side effects, it's safe to export it, and maybe some test will use it to get a sts.

File: pkg/manager/member/pd_upgrader.go [hunk 10]
```diff
@@ -16,9 +16,12 @@ package member
 import (
 	"fmt"
 
+	mngerutils "github.com/pingcap/tidb-operator/pkg/manager/utils"
+
```
Comment (DanielZhangQD) [inline pkg/manager/member/pd_upgrader.go:20]: No need to add an empty line here.

File: pkg/manager/utils/statefulset_builder.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	b.prototype.Annotations = util.CombineStringMap(b.prototype.Annotations, annos)
+}
+
+func (b *PodTemplateSpecBuilder) RunInHostNetwork() {
+	b.prototype.Spec.HostNetwork = true
+	b.prototype.Spec.DNSPolicy = corev1.DNSClusterFirstWithHostNet
+	for _, container := range b.prototype.Spec.Containers {
+		container.Env = util.AppendEnv(container.Env, []corev1.EnvVar{
+			{
+				Name: "POD_NAME",
```
Comment (DanielZhangQD) [inline pkg/manager/utils/statefulset_builder.go:104]: Why only add this env when using hostNetwork?
This is used as the advertise address, right?

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:39]: No need to distinguish, actually, we should only use the same key for all the components.

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+				},
+			},
+			Selector:                 labels,
+			PublishNotReadyAddresses: true,
+		},
+	}
+}
+
+func GenerateNGMonitoringStartScript(tngm *v1alpha1.TidbNGMonitoring) (string, error) {
+	if len(tngm.Spec.Clusters) < 1 {
```
Comment (DanielZhangQD) [inline pkg/manager/tidbngmonitoring/ng_monitoring_manager.go:463]: This should be validated before the sync starts.

File: pkg/manager/member/pd_upgrader.go (hunk 10 above)
Comment (DanielZhangQD) [reply pkg/manager/member/pd_upgrader.go:20]: Please also check the other files.

File: pkg/manager/utils/statefulset_builder.go (hunk 11 above)
Comment (KanShiori) [reply pkg/manager/utils/statefulset_builder.go:104]: yes, advertise address contains pod name

Comment (DanielZhangQD) [reply pkg/manager/utils/statefulset_builder.go:104]: So why add the env here, it should be included by default.

Comment (KanShiori) [reply pkg/manager/utils/statefulset_builder.go:104]: copy it to have same code with other managers, ngm manager will add this env in `GenerateStatefulSet`

File: pkg/manager/member/pd_upgrader.go (hunk 10 above)
Comment (KanShiori) [reply pkg/manager/member/pd_upgrader.go]: done

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (KanShiori) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go]: done

File: pkg/manager/tidbngmonitoring/ng_monitoring_manager.go (hunk 12 above)
Comment (KanShiori) [reply pkg/manager/tidbngmonitoring/ng_monitoring_manager.go]: done

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**